### ISSUE TITLE
CRD too long error with banks-vault operator helm chart 1.11 

### ISSUE THREAD

**Issue by davidgibbons** [banzaicloud/bank-vaults#1242]
> Bugs should be filed for issues encountered whilst operating bank-vaults.
> You should first attempt to resolve your issues through the community support
> channels, e.g. Slack, in order to rule out individual configuration errors.
> Please provide as much detail as possible.

**Describe the bug**:
Using the latest vault operator helm chart throws a CRD size error when deployed via argocd:
`CustomResourceDefinition.apiextensions.k8s.io "vaults.vault.banzaicloud.com" is invalid: metadata.annotations: Too long: must have at most 262144 bytes`

**Expected behaviour**:
Deploying the chart via argocd should work

**Steps to reproduce the bug**:
Attempt to add vault-operator using helm chart 1.11.0 to argocd results in the above error on sync. It appears the CRD is too large for a `kubectl apply` which is how argocd applies the resulting files.

**Additional context**:
Reverting to the 1.10 helm chart resolves the problem and allows us to deploy the vault operator chart.

**Environment details**:
- Kubernetes version (e.g. v1.10.2): 1.17.11
- Cloud-provider/provisioner (e.g. AKS, GKE, EKS, PKE etc):  EKS
- bank-vaults version (e.g. 0.4.17): banks vault operator 1.11
- Install method (e.g. helm or static manifests): helm
- Logs from the misbehaving component (and any other relevant logs):
- Resource definition (possibly in YAML format) that caused the issue, without sensitive data:

/kind bug

**Comment by elihai-risk** [comment]
having the same issue , the only to create this crd is `kubectl create -f crd.yaml`, but if you work via ArgoCD didn't find work around , hope they will have a fix for that

**Comment by elihai-risk** [comment]
@bonifaido  maybe you can fix it?

**Comment by bonifaido** [comment]
I'm not sure what could be a fix from our side here, I think this is an ArgoCD issue, namely https://github.com/argoproj/argo-cd/issues/2730.

What I can suggest is to create a copy from the chart which doesn't have an OpenAPI schema in the CRD just as it was previously and users of that version will lose all the validation that it offers.

**Comment by elihai-risk** [comment]
@is not argocd issue its k8s api limitation, try to run  `kubectl apply -f crd.yaml` and you will get this error

**Comment by bonifaido** [comment]
It is an issue in ArgoCD, because it builds on a k8s limitation. I just checked it is quite old, and unresolved since 2+ years, quite sad.

Since the operator Helm chart doesn't change too much (virtually never) except the CRD definition in it, which you can't use anyway I would suggest sticking to the operator chart version `1.10` which has no CRD schema in it, and set the image version to the most recent vault-operator image in the `values.yaml` file.

**Comment by elihai-risk** [comment]
i ran it this command ` kubectl apply -f crd.yaml` and i get this issue without ArgoCD .
any way i'm using chart 1.11.0 but i revert to this commit #910 only. and its work.
i suggest to you to try to apply your CRD with schema  and you will get this error. (btw i'm using k8s 1.18)

**Comment by davidgibbons** [comment]
I guess my hope was that there was some way to break the schema up into multiple parts that aren't too large for the kubernetes api.  This is honestly part of kube I haven't interacted with that much so maybe I'm completely off on this.

**Comment by jkirkham-ratehub** [comment]
I'm seeing this (metadata.annotations: Too long: must have at most 262144 characters) and the following issue on upgrade:
`spec.preserveUnknownFields: Invalid value: true: must be false in order to use defaults in the schema`

I'm seeing this on kubernetes 1.16 (also using ArgoCD).

**Comment by jkirkham-ratehub** [comment]
A slight modification to the CRD generator -- suppress the "description" fields can significantly reduce the size of the CRD:
```
diff --git a/main-targets.mk b/main-targets.mk
index a4113c7f..6eacdfeb 100644
--- a/main-targets.mk
+++ b/main-targets.mk
@@ -147,7 +147,7 @@ bin/controller-gen: bin/controller-gen-${CONTROLLER_GEN_VERSION}

 .PHONY: generate-crds
 generate-crds: bin/controller-gen ## Regenerate CRDs in the Helm chart and examples
-       bin/controller-gen crd paths=./operator/... output:crd:artifacts:config=./operator/deploy/
+       bin/controller-gen crd:maxDescLen=0 paths=./operator/... output:crd:artifacts:config=./operator/deploy/
        cp operator/deploy/vault.banzaicloud.com_vaults.yaml charts/vault-operator/crds/crd.yaml
        cp operator/deploy/vault.banzaicloud.com_vaults.yaml operator/deploy/crd.yaml
        rm operator/deploy/vault.banzaicloud.com_vaults.yaml
```
when I do this it seems the crd.yaml is small enough the length error goes away.

My other error appears to be something else.  I'll just have to review the CR instance to see if there are invalid fields.

**Comment by jkirkham-ratehub** [comment]
adding:
`    preserveUnknownFields: false`
to the CRD spec resolves my second issue.  It seems when it isn't specified it defaults to true.

**Comment by bonifaido** [comment]
Let me rerelease the chart without descriptions and preserveUnknownFields in the CRD inside.

**Comment by bonifaido** [comment]
Please check chart/vault-operator/1.11.1 for the reduced CRD size.

**Comment by jkirkham-ratehub** [comment]
Thanks Nándor,
OK, I've tested the vault-operator chart v1.11.1
The CRD with description fields stripped out does work, but it is still missing the `preserveUnknownFields: false` so ArgoCD failed to sync (apply) the update.

If I edit a local copy of the crd.yaml and add `preserveUnknownFields: false` (directly below "spec") and apply it then the CRD gets updated successfully.  After this ArgoCD is able to apply the real CRD from the chart in its next sync cycle.  In the end the CRD *without* `preserveUnknownFields: false` is applied and working.

I used `yq` (v4.3) to make the update in place before doing a kubectl apply:
`yq eval -i '.spec.preserveUnknownFields = false' crd.yaml`
(yq v3 would work similarly but has a very different CLI.)

I suspect this is an odd side-effect of upgrading from `apiextensions.k8s.io/v1beta1` to `apiextensions.k8s.io/v1`.  I don't think this will ever be an issue with a clean install of the chart.
It is not ideal but I think my work-around is good enough.
